### PR TITLE
fix(profiling): prepare heap for `gc_mem_caches()`

### PR DESCRIPTION
**Pull request by realFlowControl**
### Description

The PHP userland function `gc_mem_caches()` just forwards calls to `zend_mm_gc()` which will do exactly nothing and return `int(0)` in case a custom allocator is installed.
This PR makes sure we intercept userland calls to `gc_mem_caches()` and prepare the heap for such a call.

### Readiness checklist
- [x] (only for Members) Changelog has been added to the release document.
- [x] Tests added for this feature/bug.

### Reviewer checklist
- [x] Appropriate labels assigned.
- [x] Milestone is set.
- [x] Changelog has been added to the release document. For community contributors the reviewer is in charge of this task.

### CODE REVIEW

File: profiling/src/php_ffi.c [hunk 1]
```diff
@@ -75,6 +75,12 @@ void datadog_php_profiling_copy_string_view_into_zval(zval *dest, zai_string_vie
     }
 }
 
+void datadog_php_profiling_copy_long_into_zval(zval *dest, long num) {
```
Comment (morrisonlevi) [inline profiling/src/php_ffi.c:78]: Can we use the newer `ddog_php_prof_` prefix instead of `datadog_php_profiling_`?

File: profiling/src/lib.rs [hunk 2]
```diff
@@ -1045,6 +1061,43 @@ unsafe fn restore_zend_heap(heap: *mut zend::_zend_mm_heap, custom_heap: c_int)
...
 }
 
+#[cfg(feature = "allocation_profiling")]
+unsafe extern "C" fn datadog_allocation_profiling_gc_mem_caches(
+    execute_data: *mut zend::zend_execute_data,
+    return_value: *mut zend::zval,
+) {
+    // Better safe than sorry: `gc_mem_caches()` is a Zend core function which
+    // always exists
+    if GC_MEM_CACHES_HANDLER == None {
```
Comment (bwoebi) [inline profiling/src/lib.rs:1071]: wouldn't something like `let Some(func) = GC_MEM_CACHES_HANDLER else {` be much more idiomatic than having Safety annotations below?

File: profiling/src/php_ffi.c [hunk 3]
```diff
@@ -75,6 +75,12 @@ void datadog_php_profiling_copy_string_view_into_zval(zval *dest, zai_string_vie
     }
 }
 
+void ddog_php_prof_copy_long_into_zval(zval *dest, long num) {
+    ZEND_ASSERT(dest);
+    ZVAL_LONG(dest, num);
+    return;
+}
```
Comment (morrisonlevi) [inline profiling/src/php_ffi.c:82]: I suppose we could add a `ZEND_ASSERT` to be sure `dest` is undefined or a value which doesn't need refcounts nor dtor'd?

File: profiling/tests/phpt/gc_mem_caches_01.phpt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+function should actually cleanup some memory and return the amount in bytes.
+--SKIPIF--
+<?php
+if (!extension_loaded('datadog-profiling'))
+    echo "skip: test requires Datadog Continuous Profiler\n";
+?>
+--ENV--
+DD_PROFILING_ENABLED=yes
+DD_PROFILING_EXPERIMENTAL_CPU_TIME_ENABLED=yes
+DD_PROFILING_EXPERIMENTAL_ALLOCATION_ENABLED=yes
```
Comment (morrisonlevi) [inline profiling/tests/phpt/gc_mem_caches_01.phpt:16]: Probably add `DD_PROFILING_LOG_LEVEL=off` to this?

**Review by morrisonlevi — APPROVED** [review]

File: profiling/tests/phpt/gc_mem_caches_01.phpt (hunk 4 above)
Comment (realFlowControl) [reply profiling/tests/phpt/gc_mem_caches_01.phpt:16]: See #1876

**Pull request merged**